### PR TITLE
New Run() method for framework::Executor

### LINKED ISSUE #7610: How about implement another Run() in framework::Executor
Suggested by @qingqing01 , we may implement another `Run()` in [paddle/framework/executor.h](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/executor.h) which do the same things as the `Run()` of Python `Executor`.

https://github.com/PaddlePaddle/Paddle/blob/144854d2e86efce8858b34518c1ee5f4b48d9407/python/paddle/v2/fluid/executor.py#L125-L132

The interface maybe like and will do following things:
```cpp
void Run(const ProgramDesc& program,
         Scope* scope,
         std::map<std::string, Tensor>& feeds,
         std::map<std::string, Tensor>& fetchs,
         std::string& feed_var_name="feed",
         std::string& fetch_var_name="fetch") {
  1. clone the program
  2. prepend feed_op
  3. set feed variables
  4. append fetch_op
  5. call Run(const ProgramDesc&, Scope*, int, bool create_local_scope = true, bool create_vars = true)
  6. get fetch variables 
}
```

So that we can directly `Run` both in Python API and C++ API.

### PR DESCRIPTION
**Pull request by kexinzhao**
Fix #7610 

The API is a little bit different than in #7610. I encountered memory issue when using `std::map<std::string, LoDTensor>`.  So I changed to pointer instead.

```cpp
  void Run(const ProgramDesc& program, Scope* scope,
           std::map<std::string, const LoDTensor*>& feed_targets,
           std::map<std::string, LoDTensor*>& fetch_targets,
           const std::string& feed_holder_name = "feed",
           const std::string& fetch_holder_name = "fetch");
```

### CODE REVIEW

File: paddle/framework/executor.cc [hunk 1]
```diff
@@ -149,5 +150,169 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
...
+    fetch_holder = global_block->Var(fetch_holder_name);
+    fetch_holder->SetType(proto::VarDesc::FETCH_LIST);
+    fetch_holder->SetPersistable(true);
+  }
+
+  if (!has_feed_operators(global_block, feed_targets, feed_holder_name)) {
+    int i = 0;
+    for (auto& feed_target : feed_targets) {
+      std::string var_name = feed_target.first;
+      LOG(INFO) << "feed target's name: " << var_name;
```
Comment (Xreki) [inline paddle/framework/executor.cc]: LOG(INFO) -> VLOG(3)

File: paddle/framework/executor.cc [hunk 2]
```diff
@@ -149,5 +150,169 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
   }
 }
 
+// Return false if the block does not have any feed operators.
```
[the remark this answers is not in the document]
Comment (kexinzhao) [reply paddle/framework/executor.cc]: Done

File: paddle/framework/executor.cc [hunk 3]
```diff
@@ -149,5 +150,169 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
...
+// If some feed operators have been prepended to the block, check that
+// the info contained in these feed operators matches the feed_targets
+// and feed_holder_name. Raise exception when any mismatch is found.
+// Return true when the block has feed operators with matching info.
+static bool has_feed_operators(
+    BlockDesc* block, std::map<std::string, const LoDTensor*>& feed_targets,
+    const std::string& feed_holder_name) {
+  size_t feed_count = 0;
+  for (auto* op : block->AllOps()) {
+    if (op->Type() == "feed") {
```
[the remark this answers is not in the document]
Comment (kexinzhao) [reply paddle/framework/executor.cc]: Done

File: paddle/framework/executor.cc [hunk 4]
```diff
@@ -149,5 +150,169 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
...
+      feed_count++;
+      PADDLE_ENFORCE(op->Input("X")[0] == feed_holder_name,
```
[the remark this answers is not in the document]
Comment (kexinzhao) [reply paddle/framework/executor.cc]: Done

File: paddle/framework/executor.cc [hunk 5]
```diff
@@ -149,5 +150,169 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
...
+                     "Input to feed op should be '%s'", feed_holder_name);
+      std::string feed_target_name = op->Output("Out")[0];
+      PADDLE_ENFORCE(
+          feed_targets.find(feed_target_name) != feed_targets.end(),
+          "Feed operator output name '%s' cannot be found in 'feed_targets'",
+          feed_target_name);
+      PADDLE_ENFORCE(op->GetAttr("col").type() == typeid(int),
```
[the remark this answers is not in the document]
Comment (kexinzhao) [reply paddle/framework/executor.cc]: Removed this.

File: paddle/framework/executor.cc [hunk 6]
```diff
@@ -149,5 +150,169 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
...
+                     "Attribute type of 'col' should be int");
+    } else {
+      break;
+    }
+  }
+
+  if (feed_count > 0) {
+    PADDLE_ENFORCE(feed_count == feed_targets.size(),
```
[the remark this answers is not in the document]
Comment (kexinzhao) [reply paddle/framework/executor.cc]: Done

File: paddle/framework/executor.cc [hunk 7]
```diff
@@ -149,5 +150,169 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
...
+// If some fetch operators have been appended to the block, check that
+// the info contained in these fetch operators matches the fetch_targets
+// and fetch_holder_name. Raise exception when any mismatch is found.
+// Return true when the block has fetch operators with matching info.
+static bool has_fetch_operators(
+    BlockDesc* block, std::map<std::string, LoDTensor*>& fetch_targets,
+    const std::string& fetch_holder_name) {
+  size_t fetch_count = 0;
+  for (auto* op : block->AllOps()) {
+    if (op->Type() == "fetch") {
```
[the remark this answers is not in the document]
Comment (kexinzhao) [reply paddle/framework/executor.cc]: Done

File: paddle/framework/executor.cc [hunk 8]
```diff
@@ -149,5 +150,169 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
...
+    if (var->GetType() == proto::VarDesc::FEED_MINIBATCH) {
+      PADDLE_ENFORCE(var->Name() == feed_holder_name,
+                     "'feed_holder_name' should match the program desc");
+      feed_holder = var;
+    } else if (var->GetType() == proto::VarDesc::FETCH_LIST) {
+      PADDLE_ENFORCE(var->Name() == fetch_holder_name,
+                     "'fetch_holder_name' should match the program desc");
+      fetch_holder = var;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (kexinzhao) [reply paddle/framework/executor.cc]: Great point! Have removed these codes and move the check of feed/fetch_holder_name in has_feed/fetch_operator().

File: paddle/framework/executor.cc (hunk 1 above)
Comment (kexinzhao) [reply paddle/framework/executor.cc]: Done

File: paddle/framework/executor.cc [hunk 9]
```diff
@@ -149,5 +150,168 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
...
+    PADDLE_ENFORCE_EQ(var->GetType(), proto::VarDesc::FETCH_LIST,
+                      "'%s' variable should be 'FETCH_LIST' type",
+                      fetch_holder_name);
+  }
+
+  return fetch_count > 0;
+}
+
+void Executor::Run(const ProgramDesc& program, Scope* scope,
+                   std::map<std::string, const LoDTensor*>& feed_targets,
```
Comment (Xreki) [inline paddle/framework/executor.cc]: Can we use `const LoDTensor&` here?

File: paddle/framework/executor.cc [hunk 10]
```diff
@@ -149,5 +150,168 @@ void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
...
+
+  // map the data of feed_targets to feed_holder
+  for (auto* op : global_block->AllOps()) {
+    if (op->Type() == kFeedOpType) {
+      std::string feed_target_name = op->Output("Out")[0];
+      int idx = boost::get<int>(op->GetAttr("col"));
+      SetFeedVariable(scope, *feed_targets[feed_target_name], feed_holder_name,
+                      idx);
+    } else {
+      break;
```
Comment (Xreki) [inline paddle/framework/executor.cc]: Remove the else branch.

File: paddle/framework/executor.cc (hunk 9 above)
Comment (Xreki) [reply paddle/framework/executor.cc]: Reminded by @kexinzhao , it is impossible to use reference in `std::map`, https://stackoverflow.com/questions/4239253/c-is-it-possible-to-use-a-reference-as-the-value-in-a-map.

**Comment by Xreki** [comment]
@sidgoyal78 we can merge the PR first so that it will not block your work. We can fix the minor code aspects in next PR.

**Pull request merged**